### PR TITLE
Tx Size Search for Depth_2

### PR DESCRIPTION
**Pull request by hguermaz**
## Description

Tx Size Search for Depth_2.
Tx Size Search for Block Copy.

## Author(s)

@hguermaz 

## Type of change

Enhancement.

## Tests and performance
Data generated on the full objective-1-fast data set in context in enc-mode 0. 

| Speed Deviation| BD-Rate Deviation|
| -- | --
| -9.75% | -0.343%|

### CODE REVIEW

File: Source/Lib/Encoder/Codec/EbProductCodingLoop.c [hunk 1]
```diff
@@ -4409,73 +4592,205 @@ void perform_tx_partitioning(ModeDecisionCandidateBuffer *candidate_buffer,
...
+                    int16_t *dst = &(((int16_t *)context_ptr->scratch_candidate_buffer->residual_ptr
+                                          ->buffer_y)[block_index]);
+
+                    for (int i = 0; i < context_ptr->blk_geom->bheight; i++) {
+                        memcpy(dst, src, context_ptr->blk_geom->bwidth << 1);
+                        src += candidate_buffer->residual_ptr->stride_y;
+                        dst += context_ptr->scratch_candidate_buffer->residual_ptr->stride_y;
+                    }
+                }
+            }
```
Comment (mafonsoN) [inline Source/Lib/Encoder/Codec/EbProductCodingLoop.c]: I believe these copy operations could have been done in a function to make it clearer and to reduce the amount of code for a single function.

File: Source/Lib/Common/Codec/EbCommonUtils.h [hunk 2]
```diff
@@ -97,7 +97,40 @@ static INLINE int get_segdata(SegmentationParams *seg, int segment_id,
 static const PredictionMode fimode_to_intradir[FILTER_INTRA_MODES] = {
         DC_PRED, V_PRED, H_PRED, D157_PRED, DC_PRED};
 
-
+#if TXS_DEPTH_2
+static AOM_FORCE_INLINE int get_br_ctx(const uint8_t *const levels,
```
Comment (fouedb) [inline Source/Lib/Common/Codec/EbCommonUtils.h]: Add a description of the function.

File: Source/Lib/Encoder/Codec/EbProductCodingLoop.c [hunk 3]
```diff
@@ -4712,15 +5114,22 @@ void full_loop_core(PictureControlSet *pcs_ptr, SuperBlock *sb_ptr, BlkStruct *b
...
     candidate_ptr->transform_type[1] = DCT_DCT;
     candidate_ptr->transform_type[2] = DCT_DCT;
     candidate_ptr->transform_type[3] = DCT_DCT;
-
+#endif
     uint8_t start_tx_depth = 0;
     uint8_t end_tx_depth = 0;
-
+#if TXS_DEPTH_2
+    if (context_ptr->md_tx_size_search_mode == 0) {
```
[the remark this answers is not in the document]
Comment (hguermaz) [reply Source/Lib/Encoder/Codec/EbProductCodingLoop.c]: Yes. md_tx_size_search_mode controls Tx Size search per md_stage, and when 0 Tx Size search is shut.

File: Source/Lib/Encoder/Codec/EbProductCodingLoop.c (hunk 1 above)
Comment (hguermaz) [reply Source/Lib/Encoder/Codec/EbProductCodingLoop.c]: I agree. Done. 
- init_tx_candidate_buffer() to prepare tx buffers
- update_tx_candidate_buffer() to construct the final candidate buffer

File: Source/Lib/Encoder/Codec/EbProductCodingLoop.c [hunk 4]
```diff
@@ -4409,73 +4592,205 @@ void perform_tx_partitioning(ModeDecisionCandidateBuffer *candidate_buffer,
...
+                    uint16_t *dst = &(((uint16_t *)context_ptr->scratch_candidate_buffer
+                                           ->prediction_ptr->buffer_y)[block_index]);
+
+                    for (int i = 0; i < context_ptr->blk_geom->bheight; i++) {
+                        memcpy(dst, src, context_ptr->blk_geom->bwidth * sizeof(uint16_t));
+                        src += candidate_buffer->prediction_ptr->stride_y;
+                        dst += context_ptr->scratch_candidate_buffer->prediction_ptr->stride_y;
+                    }
+                }
+
```
[the remark this answers is not in the document]
Comment (hguermaz) [reply Source/Lib/Encoder/Codec/EbProductCodingLoop.c]: Done.

File: Source/Lib/Encoder/Codec/EbProductCodingLoop.c [hunk 5]
```diff
@@ -4409,73 +4592,205 @@ void perform_tx_partitioning(ModeDecisionCandidateBuffer *candidate_buffer,
...
+                                          ->buffer_y)[block_index]);
+
+                    for (int i = 0; i < context_ptr->blk_geom->bheight; i++) {
+                        memcpy(dst, src, context_ptr->blk_geom->bwidth << 1);
+                        src += candidate_buffer->residual_ptr->stride_y;
+                        dst += context_ptr->scratch_candidate_buffer->residual_ptr->stride_y;
+                    }
+                }
+            } else
+
```
[the remark this answers is not in the document]
Comment (hguermaz) [reply Source/Lib/Encoder/Codec/EbProductCodingLoop.c]: Done.

File: Source/Lib/Encoder/Codec/EbRateDistortionCost.c [hunk 6]
```diff
@@ -288,6 +288,26 @@ void eb_av1_update_eob_context(int eob, TX_SIZE tx_size, TxClass tx_class, Plane
         if (allow_update_cdf) update_cdf(ec_ctx->eob_extra_cdf[txs_ctx][plane][eob_ctx], bit, 2);
     }
 }
+#if TXS_DEPTH_2
+static int get_eob_cost(int eob, const LvMapEobCost *txb_eob_costs,
```
[the remark this answers is not in the document]
Comment (hguermaz) [reply Source/Lib/Encoder/Codec/EbRateDistortionCost.c]: Done.

**Comment by hassount** [comment]
Validated that 10bit carries similar gains with slightly higher speed costs.

**Pull request merged**